### ISSUE TITLE
Datastore indexes not found

### ISSUE THREAD

**Issue by triandco** [verida/datastore#12]
# Description
I ran into this issue when I was working on pagination for my app using a custom schema. Unfortunately, when trying to fetch documents from datastore using ```getMany({},{ limit:10, raw: true})``` instead of receiving a bookmark, I got a "no indexes found". Not sure if there is any step I missed 😅.

# Reproduction
1. Create a verida app with "ethr" as the provider
2. Create datastore with a custom schema:

```json
{
    "$id": "http://localhost:8080/schemata/schemata/matter/schema.json",
    "name": "http://localhost:8080/schemata/schemata/matter",
    "title": "Animal",
    "pluralTitle": "Animals",
    "database": {
        "name": "Animal",
        "indexes": {
            "name": [
                "name"
            ],
            "id": [
                "id"
            ]
        }
    },
    "allOf": [
        {
            "$ref": "http://localhost:8080/schemata/base/schema.json"
        },
        {
            "type": "object",
            "properties": {
                "id": {
                    "type": "string"
                },
                "name": {
                    "type": "string"
                }
            },
            "required": [
                "id",
                "name"
            ],
            "$schema": "http://json-schema.org/draft-07/schema#"
        }
    ]
}

```
3. Create a few documents using the ```datastore.save()``` method
4. Attempt to fetch documents using ```getMany({},{ limit:2, raw:true})```

# Expectation
Bookmark property is available in result for pagination

# Actual
Instead of a bookmark, we receive a warning: "No matching index found, create an index to optimize query time"

# Repository
You can find [the code here](https://github.com/triandco/datastore-indexes). 

---

Much appreciated!

**Comment by tahpot** [comment]
Thanks again for the quality / repeatable issue :)

This turned out to be a curly one!

The good news is you were doing everything correctly. The bad news is, it's complicated.

TLDR; [See pagination with Mango queries](https://pouchdb.com/guides/mango-queries.html#pagination)

The [Verida documentation / pagination section](https://docs.datastore.verida.io/#/data?id=pagination) references the `bookmark` property in the response as you correctly identified and used. However, on closer inspection it appears [PouchDB doesn't support this property](https://github.com/pouchdb/pouchdb/issues/7124). (** See sidenote below)

There is an excellent blog post discussing [pagination in PouchDB on the official website](https://pouchdb.com/2014/04/14/pagination-strategies-with-pouchdb.html). However it makes use of `allDocs()` which isn't going to give you the sort / filter capability you require.

Instead, you are better off [using pagination with Mango queries](https://pouchdb.com/guides/mango-queries.html#pagination). In your example, this would mean the following (untested):

```
let lastId = null
const currentPageResults = await datastore.getMany({
  _id: {$gt: lastId}
}, {
  limit: 10,
  sort:['name']
})
```

A couple of other things I noticed in the code example about the schema that may be helpful to highlight:

1. The schema defines a custom `id` property. Instead you are better off using the inbuilt `_id` property as it will automatically be created as an index and also enforce uniqueness.
2. The schema definition references a copy of the `base` Verida schema. You can alternatively reference the URL of the base Verida schema and your custom schema will inherit it. ie: replace `{ "$ref": "http://localhost:8080/schemata/base/schema.json" },` with `{ "$ref": "https://schemas.verida.io/base/schema.json" },`

---

** What makes this more confusing is that Verida has two types of database instances. It has native `public` databases that are unencrypted. These databases talk directly to CouchDB which supports the `bookmark` property so it is returned. Verida also has native `encrypted` databases that connect to an encrypted CouchDB instance, create a local PouchDB instance and populate it with the decrypted data. When you call `find()` you are actually querying the local copy that has been decrypted and sync'd. As such, you don't receive the `bookmark` property (but I'm guessing you would if you were talking to an unencrypted database).

**Comment by tahpot** [comment]
@Dickson-Mwendia Could you look at updating our Datastore documentation.

- Perhaps we replace the pagination section to not mention `bookmark` and instead mention the strategy above?
- Make the schema documentation clearer about how `_id` primary keys work and referencing the base Verida schemas

**Comment by Dickson-Mwendia** [comment]
@tahpot sure, I'll definitely look into this and update our docs. 

Issue tracked in work item - https://www.notion.so/verida/Update-pagination-section-of-the-datastore-documentation-decdb2bdb26d475d9170e13f84fd83bd

**Comment by triandco** [comment]
Thank you @tahpot. 

I'll try the Mango pagination queries strategy. 

Thanks for the hint with the _id, I'll use it instead of defining a new one  👍.

Re the base schema, I swapped it to ```https://schemas.verida.io/base/schema.json``` at it works fine. At the moment, ```https://schemas.verida.io/base/schema.json``` is currently not reachable. Does it mean the URL is only used as an id for internal reference rather than trying to fetch the json from ```schemas.verida.io```?

**Comment by tahpot** [comment]
>Re the base schema, I swapped it to https://schemas.verida.io/base/schema.json at it works fine. At the moment, https://schemas.verida.io/base/schema.json is currently not reachable. Does it mean the URL is only used as an id for internal reference rather than trying to fetch the json from schemas.verida.io?

Ah yes.

So `https://schemas.verida.io/base/schema.json` is not reachable, however the `datastore` supports configuration to "re-map" the schemas to other locations to assist with development. As such, the current `testnet` configuration remaps the base schemas (schemas.verida.io) to the working URL (`https://schemas.testnet.verida.io/base/schema.json`).

This is important as the schema URL is embedded in the records themselves and can cause a lot of confusion if the wrong schema is stored.

In the future we may migrate to a more content addressable schema approach (ie: IPFS), however hosting schemas via HTTP is much simpler and flexible in this early stage.
